src/AddAtomicMutex.cpp

```diff
@@ -0,0 +1,344 @@
+                Expr e = let_bindings.get(op->name);
+                FindLoad finder;
+                if (finder.find_load(e, store_names)) {
+                    found = true;
+                }
+            }
+        }
+    }
+
+    void visit(const Store *op) override {
+        include(op->predicate);
+        inside_store = true;
+        include(op->value);
+        inside_store = false;
+        include(op->index);
+    }
+
+    bool inside_store;
+    set<string> store_names;
+    Scope<Expr> let_bindings;
+    bool found = false;
+};
+
+class CollectStoreNames : public IRGraphVisitor {
+public:
+    using IRGraphVisitor::visit;
+    
+    void visit(const Store *op) override {
+        include(op->predicate);
+        include(op->value);
+        include(op->index);
+        store_names.insert(op->name);
+    }
+
+    set<string> store_names;
+};
+
+class RemoveUnnecessaryMutexUse : public IRMutator {
+public:
+    using IRMutator::visit;
+
+    Stmt visit(const Atomic *op) override {
+        // Collect the names of all Store nodes inside.
+        CollectStoreNames collector;
+        op->body.accept(&collector);
+        // Search for let bindings that access the producers.
+        FindAtomicLetBindings finder(collector.store_names);
+        op->body.accept(&finder);
+        if (finder.found) {
+            // Can't remove mutex lock. Leave the Stmt as is.
+            return IRMutator::visit(op);
+        } else {
+            remove_mutex_lock_names.insert(op->mutex_name);
+            Stmt body = mutate(op->body);
+            return Atomic::make(op->producer_name,
+                                "",
+                                std::move(body));
+        }
+    }
+
+    std::set<string> remove_mutex_lock_names;
+};
+
+class FindStoreInAtomicMutex : public IRGraphVisitor {
+public:
+    using IRGraphVisitor::visit;
+
+    FindStoreInAtomicMutex(const std::set<std::string> &store_names)
+        : store_names(store_names) {}
+
+    void visit(const Atomic *op) override {
+        if (!found && !op->mutex_name.empty()) {
+            ScopedValue<bool> old_in_atomic_mutex(in_atomic_mutex, true);
+            include(op->body);
+            if (found) {
+                // We found a Store inside Atomic with matching name,
+                // record the mutex information.
+                producer_name = op->producer_name;
+                mutex_name = op->mutex_name;
+            }
+        } else {
+            include(op->body);
+        }
+    }
+
+    void visit(const Store *op) override {
+        if (in_atomic_mutex) {
+            if (store_names.find(op->name) != store_names.end()) {
+                found = true;
+            }
+        }
+        IRGraphVisitor::visit(op);
+    }
+
+    bool in_atomic_mutex = false;
+    const std::set<std::string> &store_names;
+    bool found = false;
+    std::string producer_name;
+    std::string mutex_name;
+};
+
+class FindStoreIndex : public IRGraphVisitor {
```

Comment:
What does this class even do?

Reply:
Fixed. This is a bug. Added comments to each class regarding what they do.
The reason why this didn't trigger an error in the tests is because this bug makes all the threads access mutex at location 0, which means it just serializes all the atomic mutex access.